src/Connection/DefaultConnectionPool.php

```diff
@@ -82,141 +95,171 @@ public function getStream(Request $request, CancellationToken $cancellation): Pr
+                "Connection to '%s' timed out, took longer than " . $request->getTcpConnectTimeout() . ' ms',
+                $authority
+            )); // don't pass $e
+        }
+
+        if (!$isHttps) {
+            return new Http1Connection($socket);
+        }
+
+        try {
+            $tlsState = $socket->getTlsState();
+            if ($tlsState === EncryptableSocket::TLS_STATE_DISABLED) {
+                $tlsCancellationToken = new CombinedCancellationToken(
+                    $cancellation,
+                    new TimeoutCancellationToken($request->getTlsHandshakeTimeout())
+                );
+                yield $socket->setupTls($tlsCancellationToken);
+            } elseif ($tlsState !== EncryptableSocket::TLS_STATE_ENABLED) {
+                $socket->close();
+                throw new SocketException('Failed to setup TLS connection, connection was in an unexpected TLS state (' . $tlsState . ')');
+            }
+        } catch (StreamException $exception) {
+            $socket->close();
+            throw new SocketException(\sprintf(
+                "Connection to '%s' closed during TLS handshake",
+                $authority
+            ), 0, $exception);
+        } catch (CancelledException $e) {
+            $socket->close();
+
+            // In case of a user cancellation request, throw the expected exception
+            $cancellation->throwIfRequested();
+
+            // Otherwise we ran into a timeout of our TimeoutCancellationToken
+            throw new TimeoutException(\sprintf(
+                "TLS handshake with '%s' @ '%s' timed out, took longer than " . $request->getTlsHandshakeTimeout() . ' ms',
+                $authority,
+                $socket->getRemoteAddress()->toString()
+            )); // don't pass $e
+        }
+
+        $tlsInfo = $socket->getTlsInfo();
+        if ($tlsInfo === null) {
+            throw new SocketException('Socket disconnected immediately after enabling TLS');
+        }
+
+        if ($tlsInfo->getApplicationLayerProtocol() === 'h2') {
+            $connection = new Http2Connection($socket);
+            yield $connection->initialize();
+
+            return $connection;
+        }
+
+        if (!\array_intersect($request->getProtocolVersions(), ['1.0', '1.1'])) {
+            $socket->close();
+            throw new InvalidRequestException(
+                $request,
+                'Downgrade to HTTP/1.x forbidden, but server does not support HTTP/2'
+            );
+        }
+
+        return new Http1Connection($socket);
+    }
+
+    /**
+     * @param int $timeout Number of milliseconds before the estimated connection timeout that a non-idempotent
+     *                     request should will not be sent on an existing HTTP/1.x connection, instead opening a
+     *                     new connection for the request. Default is 2000 ms.
+     */
+    public function setConnectionTimeoutDelta(int $timeout): void
```

Comment:
`setTimeoutGracePeriod`?

Reply:
Yes, much better. Should the pool be immutable, or no?

Reply:
If so, we'd have to empty `$connections` on `__clone`.